presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java

```diff
@@ -321,14 +323,341 @@ public void testHardCpuLimit()
         assertEquals(query2.getState(), QUEUED);
 
         query1.complete();
-        root.processQueuedQueries();
+        root.updateGroupsAndProcessQueuedQueries();
         assertEquals(query2.getState(), QUEUED);
 
         root.generateCpuQuota(2);
-        root.processQueuedQueries();
+        root.updateGroupsAndProcessQueuedQueries();
         assertEquals(query2.getState(), RUNNING);
     }
 
+    /**
+     * Test resource group cpu usage update by manually invoking the cpu quota regeneration and queue processing methods
+     * that are invoked periodically by the resource group manager
+     */
+    @Test(timeOut = 10_000)
+    public void testCpuUsageUpdateForRunningQuery()
+    {
+        RootInternalResourceGroup root = new RootInternalResourceGroup("root", (group, export) -> {}, directExecutor());
+        InternalResourceGroup child = root.getOrCreateSubGroup("child");
+
+        // Set the common configurations for the two groups
+        List<InternalResourceGroup> groups = ImmutableList.of(root, child);
+        groups.forEach(group -> {
+            group.setCpuQuotaGenerationMillisPerSecond((new Duration(1, SECONDS)).toMillis());
+            group.setHardCpuLimit(new Duration(1.5, SECONDS));
+            group.setSoftCpuLimit(new Duration(1.5, SECONDS));
+            group.setHardConcurrencyLimit(100);
+            group.setMaxQueuedQueries(100);
+        });
+
+        MockResourceUpdatingQueryExecution q1 = new MockResourceUpdatingQueryExecution();
+        child.run(q1);
+        assertEquals(q1.getState(), RUNNING);
+
+        // The cached cpu usage will be set to 2s (> 1.5s limit) when updateGroupsAndProcessQueuedQueries is invoked
```

Comment:
I agree that it is a better way to assert the status of resource group usage. But for this direct verification, we'd need to put a lock on root for creating a snapshot for `cachedResourceUsage`, since this object is guarded by it. I'm not sure if it's a good idea to use the root lock just for this test, since someone else may start using it in main code. What do you think?